rootfs/etc/nginx/lua/test/balancer/ewma_test.lua

```diff
@@ -96,6 +108,24 @@ describe("Balancer ewma", function()
       assert.equal("10.10.10.3:8080", peer)
       assert.are.equals(0.16240233988393523723, ngx.var.balancer_ewma_score)
     end)
+
+    it("doesn't pick the tried endpoint while retry", function()
+      local two_endpoints_backend = util.deepcopy(backend)
+      table.remove(two_endpoints_backend.endpoints, 2)
+      local two_endpoints_instance = balancer_ewma:new(two_endpoints_backend)
+
+      local peer = two_endpoints_instance:balance()
+      assert.equal("10.10.10.1:8080", peer)
+    end)
+
+    it("all the endpoints are tried, pick the one with lowest score", function()
+      local two_endpoints_backend = util.deepcopy(backend)
+      table.remove(two_endpoints_backend.endpoints, 2)
+      local two_endpoints_instance = balancer_ewma:new(two_endpoints_backend)
+
+      local peer = two_endpoints_instance:balance()
+      assert.equal("10.10.10.3:8080", peer)
+    end)
```

Comment:
I don't understand the logic in these two tests, where are you marking the endpoints as tried? I'd have expected you modify `ngx.ctx.balancer_ewma_tried_endpoints` to control these tests.

--

Also an additional test is needed to ensure `ngx.ctx.balancer_ewma_tried_endpoints` is set in balance function.

Reply:
`ngx.ctx` is shared between tests. I will change the test to use `ngx.ctx` directly so the test will be more clear.

Reply:
@ElvinEfendi 
Updated.

Reply:
I see, yeah it is better to keep the tests independent of each other.